content/blog/2022/2022-02-15-getting-started-with-spring-webflux.md

```diff
@@ -0,0 +1,566 @@
+
+```java
+    public Mono<User> postUser(User user) {
+        return webClient
+                .post()
+                .uri("http://localhost:9000/api/users")
+                .header("Authorization", "Basic MY_PASSWORD")
+                .accept(MediaType.APPLICATION_JSON)
+                .body(Mono.just(user), User.class)
+                .retrieve()
+                .bodyToMono(User.class)
+                .log()
+                .retryWhen(Retry.backoff(10, Duration.ofSeconds(2)))
+                .onErrorReturn(new User())
+                .doOnError(throwable -> log.error("Result error out for POST user", throwable))
+                .doFinally(signalType -> log.info("Result Completed for POST User: {}", signalType));
+    }
+```
+
+If we receive the response in the first attempt, then it will simply return the response back. But due to some reason, if the API faces some issue, then it will try to re-attempt for a while after which it can return some indicated data or logs to denote that it has failed due to some reason.
+
+## Alternative server used in WebFlux
+
+Traditionally, Spring MVC bundles *Tomcat* server for servlet stack applications whereas Spring WebFlux bundles *Reactor Netty* by default for reactive stack applications. *Reactor Netty* is an asynchronous event-driven network application framework built out of Netty server which provides non-blocking and backpressure-ready network engines for HTTP, TCP and UDP clients and servers. Spring WebFlux automatically configures Reactor Netty as the default server but if we want to override the default configuration, we can simply do that by defining them under server prefix.
+
+```yaml
+
+server:
+  port: 9000
+  http2:
+    enabled: true
+```
+
+We can also define the other properties in the same way that start with the `server` prefix by overriding the default server configuration.
+
+## Concurrency Model
+
+Usually, as we discussed above the traditional Spring Boot applications serve thread-per-request model which means each request to a web server can be handled by different threads allocated per request. However, most of the communications within a given thread are still blocking in nature. Hence, reactive programming came up with a different approach to place a concurrency model that can serve more requests with relatively less number of threads.
+
+Let’s say in a reactive model, a read call to any DB will not block the calling thread while data is retrieved or exchanged. The call will return a publisher immediately that others can subscribe. The subscriber can process the events later after the result has been generated or processed. The overall focus relies on re-structuring the system as some kind of *asynchronous event data stream*. 
+
+{{% image alt="Concurrency in WebFlux" src="images/posts/spring-webflux/webflux_concurrency.png" %}}
```

Comment:
There's a typo in the image. It should be 

- "A single or a few threads process all requests"
- "Thread registers callback and doesn't block"

Reply:
Updated the image.